lib/Raven/Breadcrumbs/Breadcrumb.php

```diff
@@ -0,0 +1,261 @@
+<?php
+
+/*
+ * This file is part of Raven.
+ *
+ * (c) Sentry Team
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Raven\Breadcrumbs;
+
+use Psr\Log\LogLevel;
+use Raven\Exception\InvalidArgumentException;
+
+/**
+ * This class stores all the informations about a breadcrumb.
+ *
+ * @author Stefano Arlandini <sarlandini@alice.it>
+ */
+final class Breadcrumb implements \JsonSerializable
+{
+    /**
+     * This constant defines the http breadcrumb type.
+     */
+    const TYPE_HTTP = 'http';
+
+    /**
+     * This constant defines the user breadcrumb type.
+     */
+    const TYPE_USER = 'user';
+
+    /**
+     * This constant defines the navigation breadcrumb type.
+     */
+    const TYPE_NAVIGATION = 'navigation';
+
+    /**
+     * This constant defines the error breadcrumb type.
+     */
+    const TYPE_ERROR = 'error';
+
+    /**
+     * @var string The category of the breadcrumb
+     */
+    private $category;
+
+    /**
+     * @var string The type of breadcrumb
+     */
+    private $type;
+
+    /**
+     * @var string The message of the breadcrumb
+     */
+    private $message;
+
+    /**
+     * @var string The level of the breadcrumb
+     */
+    private $level;
+
+    /**
+     * @var array The meta data of the breadcrumb
+     */
+    private $metaData;
+
+    /**
+     * @var float The timestamp of the breadcrumb
+     */
+    private $timestamp;
+
+    /**
+     * Constructor.
+     *
+     * @param string      $level    The error level of the breadcrumb
+     * @param string      $type     The type of the breadcrumb
+     * @param string      $category The category of the breadcrumb
+     * @param string|null $message  Optional text message
+     * @param array       $metaData Additional information about the breadcrumb
+     */
+    public function __construct($level, $type, $category, $message = null, array $metaData = [])
+    {
+        if (!in_array($level, self::getLevels(), true)) {
+            throw new InvalidArgumentException('The value of the $level argument must be one of the constants in Psr\Log\LogLevel.');
+        }
+
+        if (!in_array($type, self::getTypes(), true)) {
```

Comment:
this shouldn't be restricted like this as type is defined on the server and can be expanded on